pkg/util/kubernetes/apiserver/hpa_controller.go

```diff
@@ -184,9 +187,16 @@ func (h *AutoscalersController) pushToGlobalStore() error {
 }
 
 func (h *AutoscalersController) updateExternalMetrics() {
+	// We force a flush of the most up to date metrics that might remain in the batch
+	err := h.pushToGlobalStore()
+	if err != nil {
+		log.Errorf("Error while pushing external metrics to the store: %s", err)
+		return
+	}
+
```

Comment:
I think we need to lock the store here to fully prevent the race condition. What if:

```go
err := h.pushToGlobalStore()
emList, err := h.store.ListAllExternalMetricValues()
// <---- some update happens now and goes to the store ---->
// <---- batchFreq ticks and we write that update to the CM ---->
updated := h.hpaProc.UpdateExternalMetrics(emList)
err = h.store.SetExternalMetricValues(updated)
 ```

`h.store.SetExternalMetricValues(updated)` would override a legit update that happened just before.

Please let me know if I missed something.

Reply:
```suggestion

	// prevent some update to get into the store between ListAllExternalMetricValues
	// and SetExternalMetricValues otherwise it would get erased
	h.toStore.m.Lock()
	defer h.toStore.m.Unlock()

```